link-grammar/tokenize/tokenize.c

```diff
@@ -3275,11 +3274,12 @@ static X_node * build_word_expressions(Sentence sent, const Gword *w,
 		 * result is valid (in case of "\*" it means an empty dict or a dict
 		 * only with walls that are not used).  Use a null-expression
 		 * instead, to prevent an error at the caller. */
-		y = pool_alloc(sent->X_node_pool);
+		X_node * y = pool_alloc(sent->X_node_pool);
 		y->next = NULL;
 		y->exp = make_zeroary_node(sent->Exp_pool);
 	}
 
+	assert(NULL != x, "Word '%s': NULL X-node", w->subword);
```

Comment:
It was not an `assert()` by intention, as historically it happened in releases due to rare conditions (that have been fixed for the next release) and I didn't think there is a need to crash the whole program due to that.

Reply:
I did see this myself, when I had done something wrong with the dict. It was "obviously" a dict bug, so I had to fix it.  And that's why turning this into an assert seemed OK.

I can chance it ... what do you want? Make it a debug print, again?

Reply:
s/chance/change/

Reply:
I think that `assert()`  is better to be reserved for internal errors and not for definition errors external to the program, for which I think `prt_error()` + `exit(NONZERO)` is more appropriate.
Here, this condition could also happen due to a complex tokenization error that happens only on rare conditions, so a crash is not needed (as opposed to an `assert()` that validates an "impossible" condition that if it is false it means the program might have a memory trashing and all chances are lost.

Reply:
This assert is now in `tokenize/lookup-exprs.c` line 126 

I did recently hit it, due to a code bug: for the atomese backend, `exists_lookup` reported that a word was present in the dict, but the actual `dictionary_lookup_list` returned nothing.

I don't see how it is possible for dictionary errors to trigger this assert, but ... maybe. Thus, I don't know what kind of error message to print, that would make sense to the user.